augur/align.py

```diff
@@ -51,79 +38,155 @@ def run(args):
-        ref_name = None
-
-    # potentially add the reference seuqnce to the sequences
-    if ref_fname and (not ref_name):
-        if os.path.isfile(ref_fname):
-            try:
-                ref_seq = SeqIO.read(ref_fname, 'genbank' if ref_fname.split('.')[-1] in ['gb', 'genbank'] else 'fasta')
-            except:
-                print("WARNING: Cannot read reference sequence."
-                      "\n\tmake sure the file %s contains one sequence in genbank or fasta format"%ref_fname)
+        # If given an existing alignment, then add the reference sequence to this if desired (and if it is the same length)
+        if existing_aln and args.reference_sequence:
+            existing_aln_fname = args.existing_alignment + ".ref.fasta"
+            ref_seq = read_reference(args.reference_sequence)
+            if len(ref_seq) != existing_aln.get_alignment_length():
+                raise AlignmentError("ERROR: Provided existing alignment ({}bp) is not the same length as the reference sequence ({}bp)".format(alignment.get_alignment_length(), len(ref_seq)))
+            existing_aln.append(ref_seq)
+            SeqIO.write(existing_aln, existing_aln_fname, 'fasta')
+            temp_files_to_remove.append(existing_aln_fname)
+            ref_name = ref_seq.id
+        else:
+            existing_aln_fname = args.existing_alignment # may be False
+
+        ## Given >=1 sequence files, create a single file for alignment. Add in the reference
+        ## sequence if desired
+        if args.reference_sequence and not existing_aln:
+            seqs_to_align_fname = args.output+".to_align.fasta"
+            temp_files_to_remove.append(seqs_to_align_fname)
+            ref_seq = read_reference(args.reference_sequence)
+            SeqIO.write(list(seqs.values())+[ref_seq], seqs_to_align_fname, 'fasta')
+            ref_name = ref_seq.id
+        elif len(args.sequences) > 1:
+            if existing_aln:
+                seqs_to_align_fname = args.output+".new_seqs_to_align.fasta"
             else:
-                ref_name = ref_seq.id
-                seq_fname+='.ref.fasta'
-                SeqIO.write(list(seqs.values())+[ref_seq], seq_fname, 'fasta')
+                seqs_to_align_fname = args.output+".to_align.fasta"
+            SeqIO.write(list(seqs.values()), seqs_to_align_fname, 'fasta')
+            temp_files_to_remove.append(seqs_to_align_fname)
         else:
-            print("WARNING: Cannot read reference sequence."
-                  "\n\tmake sure the file %s does not exist"%ref_fname)
+            seqs_to_align_fname = args.sequences[0]
+
+        check_duplicates(existing_aln, ref_name, seqs)
+
+        # before aligning, make a copy of the data that the aligner receives as input (very useful for debugging purposes)
+        if not existing_aln:
+            copyfile(seqs_to_align_fname, args.output+".pre_aligner.fasta")
```

Comment:
these files could also be removed by default

Reply:
👍 05864fa